src/analysis/processing/qgsalgorithmtransect.cpp

```diff
@@ -0,0 +1,186 @@
+                90.0, false, 0, 360 ) );
+  addParameter( new QgsProcessingParameterEnum( QStringLiteral( "SIDE" ), QObject::tr( "Side to create the transects" ), QStringList() << QObject::tr( "Left" ) << QObject::tr( "Right" ) << QObject::tr( "Both" ), false ) );
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Transect" ), QgsProcessing::TypeVectorLine ) );
+
+}
+
+QString QgsTransectAlgorithm::shortHelpString() const
+{
+
+  return QObject::tr( "This algorithm creates transects on vertices for (multi)linestring.\n" ) +
+         QObject::tr( "A transect is a line oriented from an angle (by default perpendicular) to the input polylines (at vertices)." ) +
+         QStringLiteral( "\n\n" )  +
+         QObject::tr( "Field(s) from feature(s) are returned in the transect with these new fields:\n" ) +
+         QObject::tr( "- TR_FID: ID of the original feature\n" ) +
+         QObject::tr( "- TR_ID: ID of the transect. Each transect have an unique ID\n" ) +
+         QObject::tr( "- TR_SEGMENT: ID of the segment of the linestring\n" ) +
+         QObject::tr( "- TR_ANGLE: Angle in degrees from the original line at the vertex\n" ) +
+         QObject::tr( "- TR_LENGTH: Total length of the transect returned\n" ) +
+         QObject::tr( "- TR_ORIENT: Side of the transect (only on the left or right of the line, or both side)\n" );
+
+}
+
+QgsTransectAlgorithm *QgsTransectAlgorithm::createInstance() const
+{
+  return new QgsTransectAlgorithm();
+}
+
+QVariantMap QgsTransectAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  Side orientation = static_cast< QgsTransectAlgorithm::Side >( parameterAsInt( parameters, QStringLiteral( "SIDE" ), context ) );
+  double angle = fabs( parameterAsDouble( parameters, QStringLiteral( "ANGLE" ), context ) );
+  double length = parameterAsDouble( parameters, QStringLiteral( "LENGTH" ), context );
+
+  if ( orientation == QgsTransectAlgorithm::Both )
+    length /= 2.0;
+
+  std::unique_ptr< QgsFeatureSource > source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !source )
+    return QVariantMap();
+
+  QgsFields fields = source->fields();
+
+  fields.append( QgsField( QStringLiteral( "TR_FID" ), QVariant::Int, QString(), 20 ) );
+  fields.append( QgsField( QStringLiteral( "TR_ID" ), QVariant::Int, QString(), 20 ) );
+  fields.append( QgsField( QStringLiteral( "TR_SEGMENT" ), QVariant::Int, QString(), 20 ) );
+  fields.append( QgsField( QStringLiteral( "TR_ANGLE" ), QVariant::Double, QString(), 5, 2 ) );
+  fields.append( QgsField( QStringLiteral( "TR_LENGTH" ), QVariant::Double, QString(), 20, 6 ) );
+  fields.append( QgsField( QStringLiteral( "TR_ORIENT" ), QVariant::Int, QString(), 1 ) );
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, fields,
```

Comment:
Note - when https://github.com/qgis/QGIS/pull/5418 lands, the only thing required to keep Z/M values here is to change:

```
QgsWkbTypes::Type outWkb = QgsWkbTypes::LineString;
  if ( QgsWkbTypes::hasZ( source->wkbType()  ) )
    outputWkb = QgsWkbTypes::addZ( outputWkb );
  if ( QgsWkbTypes::hasM( source->wkbType() ) )
    outputWkb = QgsWkbTypes::addM( outputWkb );

std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, fields, outputWkb, source->sourceCrs() );
```